### PR TITLE
Adding in the powerful P combinator

### PR DESCRIPTION
**Pull request by dalefrancis88**
This combinator, once i wrapped my head around it, has become one of my go to tools since. Like `converge` it allows for some really nice/neat compositions for inputs into those `binary` functions we all love :)

### CODE REVIEW

**Comment by dalefrancis88** [comment]
@evilsoft I'm hoping to get this into the next release along with `bichain` but if there are issues, don't hesitate to release `bichain` this can wait

File: docs/src/pages/docs/functions/combinators.md [hunk 1]
```diff
@@ -309,6 +309,90 @@ with this library. It quite simply is just a function that when you pass it
 something, it returns that thing right back to you. So simple, I will leave it
 as an exercise to reason about why this is so powerful and important.
 
+#### psi
```
Comment (bennypowers) [inline docs/src/pages/docs/functions/combinators.md]: for the other combinators, we've in the past prefered descriptive names over conventional ones. (i.e. 'flip' vs 'kestrel' or 'C')

in this case, maybe we'd prefer 'on' ?

File: docs/src/pages/docs/functions/combinators.md [hunk 2]
```diff
@@ -309,6 +309,90 @@ with this library. It quite simply is just a function that when you pass it
...
+
+`crocks/combinators/psi`
+
+```haskell
+psi ::  (b -> b -> c) -> (a -> b) -> a -> a -> c
+```
+
+`psi` is a function that can be considered the sister of [`converge`](#converge).
+Where [`converge`](#converge) takes one argument an maps it through two functions,
```
Comment (bennypowers) [inline docs/src/pages/docs/functions/combinators.md]:
```suggestion
Where [`converge`](#converge) takes one argument and maps it through two functions,
```

File: docs/src/pages/docs/functions/combinators.md [hunk 3]
```diff
@@ -309,6 +309,90 @@ with this library. It quite simply is just a function that when you pass it
...
+merging the resulting values with an accumulating function, `psi` takes two
```
Comment (bennypowers) [inline docs/src/pages/docs/functions/combinators.md]:
```suggestion
merging the resulting values with an binary combination function, `psi` takes two
```

File: docs/src/pages/docs/functions/combinators.md [hunk 4]
```diff
@@ -309,6 +309,90 @@ with this library. It quite simply is just a function that when you pass it
...
+arguments and runs them through the same function individually before merging
```
Comment (bennypowers) [inline docs/src/pages/docs/functions/combinators.md]:
```suggestion
arguments and runs each through the same unary function before merging
```

File: docs/src/pages/docs/functions/combinators.md [hunk 5]
```diff
@@ -309,6 +309,90 @@ with this library. It quite simply is just a function that when you pass it
...
+them with the given accumulating function.
```
Comment (bennypowers) [inline docs/src/pages/docs/functions/combinators.md]:
```suggestion
them with the given binary combination function.
```

File: docs/src/pages/docs/functions/combinators.md (hunk 3 above)
Comment (bennypowers) [reply docs/src/pages/docs/functions/combinators.md]: `binary` is important. If there are other instances in docs where `accumulating` is preferred over `combining`, or if there is some other term, use that.

File: docs/src/pages/docs/functions/combinators.md [hunk 6]
```diff
@@ -309,6 +309,90 @@ with this library. It quite simply is just a function that when you pass it
...
+
+`psi` is often used to easily [`compose`][compose] equality checking functions or when
```
Comment (bennypowers) [inline docs/src/pages/docs/functions/combinators.md]:
```suggestion
`psi` is often used to [`compose`][compose] equality checking functions or when
```

File: docs/src/pages/docs/functions/combinators.md [hunk 7]
```diff
@@ -309,6 +309,90 @@ with this library. It quite simply is just a function that when you pass it
...
+import and from 'crocks/logic/and'
+import equals from 'crocks/pointfree/equals'
+import isNumber from 'crocks/predicates/isNumber'
+import liftA2 from 'crocks/helpers/liftA2'
+import safe from 'crocks/Maybe/safe'
+
+// isNonZero :: Number -> Boolean
+const isNonZero = x => x !== 0
+
+// isValid :: Number -> Boolean
```
Comment (bennypowers) [inline docs/src/pages/docs/functions/combinators.md]:
```suggestion:+2
// isValidDivisor :: Number -> Boolean
const isValidDivisor = and(isNumber, isNonZero) 
```

File: docs/src/pages/docs/functions/combinators.md [hunk 8]
```diff
@@ -309,6 +309,90 @@ with this library. It quite simply is just a function that when you pass it
...
+const isValid = and(isNumber, isNonZero)
+
+// divideBy :: Number -> Number -> Number
+const divideBy = x => y => y / x
+
+// safeDivide :: Number -> Number -> Maybe Number
+const safeDivide =
+  psi(liftA2(divideBy), safe(isValid))
```
Comment (bennypowers) [inline docs/src/pages/docs/functions/combinators.md]:
```suggestion
  psi(liftA2(divideBy), safe(isValidDivisor))
```

File: docs/src/pages/docs/functions/combinators.md [hunk 9]
```diff
@@ -309,6 +309,90 @@ with this library. It quite simply is just a function that when you pass it
...
+safeDivide('0.5', 21)
+//=> Nothing
+
+safeDivide(0.5, '21')
+//=> Nothing
+
+safeDivide(29, 0)
+//=> Nothing
+
+// toTitleCase :: String -> String
```
Comment (bennypowers) [inline docs/src/pages/docs/functions/combinators.md]:
```suggestion:+2
// capitalize :: String -> String
const capitalize = str => 
```

File: docs/src/pages/docs/functions/combinators.md [hunk 10]
```diff
@@ -309,6 +309,90 @@ with this library. It quite simply is just a function that when you pass it
...
+const toTitleCase = str =>
+  `${str.charAt(0).toUpperCase()}${str.slice(1)}`
+
+// join :: String -> String -> String -> String
+const join = delim => right => left =>
+  `${left}${delim}${right}`
+
+// createName :: String -> String -> String
+const createName =
+  psi(join(', '), toTitleCase)
```
Comment (bennypowers) [inline docs/src/pages/docs/functions/combinators.md]:
```suggestion
  psi(join(', '), capitalize)
```

File: docs/src/pages/docs/functions/combinators.md [hunk 11]
```diff
@@ -309,6 +309,90 @@ with this library. It quite simply is just a function that when you pass it
...
+
+createName('sara', 'smith')
+//=> Smith, Sara
+
+// toLowerCase :: String -> String
+const toLowerCase = str =>
+  str.toLowerCase()
+
+// equalsIgnoreCase :: String -> String -> Boolean
+const equalsIgnoreCase =
```
Comment (bennypowers) [inline docs/src/pages/docs/functions/combinators.md]: great example this really drives the use case home!

File: docs/src/pages/docs/functions/combinators.md (hunk 1 above)
Comment (dalefrancis88) [reply docs/src/pages/docs/functions/combinators.md]: Yeah i know `on` is what this is called in Haskell and Sanctuary but i don't find it to be that descriptive honestly, I got the original reference and name from [here](https://gist.github.com/Avaq/1f0636ec5c8d6aed2e45) we seem to have matched the names with most everything else except `thrush`

File: docs/src/pages/docs/functions/combinators.md (hunk 3 above)
Comment (dalefrancis88) [reply docs/src/pages/docs/functions/combinators.md]: So this is actually similiar to what i wrote originally, my preference is `binary` over `accumulating` however i changed this to match the phrasing of `converge` and have a common voice. 
Should we make a point of changing the docs for `converge`?

Comment (dalefrancis88) [reply docs/src/pages/docs/functions/combinators.md]: I might also drop combination from your suggestion, i'm not sure it's neccessary

File: docs/src/pages/docs/functions/combinators.md (hunk 11 above)
Comment (dalefrancis88) [reply docs/src/pages/docs/functions/combinators.md]: Yeah equality checking is really where `psi` earns it's stripes

File: docs/src/pages/docs/functions/combinators.md (hunk 1 above)
Comment (bennypowers) [reply docs/src/pages/docs/functions/combinators.md]: looks like a solid precedent. Thanks for addressing this.

File: docs/src/pages/docs/functions/combinators.md (hunk 3 above)
Comment (bennypowers) [reply docs/src/pages/docs/functions/combinators.md]: sounds good to me. To my ear, binary is clearer than accumulating, so i'd prefer to see that jargon adopted globally @evilsoft ?

Comment (dalefrancis88) [reply docs/src/pages/docs/functions/combinators.md]: Yeah i would agree 100%

**Pull request merged**